### PR TITLE
Disentangle the fetchers!

### PR DESCRIPTION
**Pull request by alalazo**
It's a while that I want to break the tight coupling between fetchers and stages, and here it is.

The main gain from this PR is that now fetch strategies are completely decoupled from stages. Their API changed a bit, and can be driven by setting a couple of paths prior to any relevant method call. The most important path to be set is:
```python
fetcher.source_dir = path
```
which in most cases is set to `stage.path` within Spack. Another relevant path attribute is `expected_archive_files`. In any case all the computations a fetcher does now should depend only on its current state (while before they were also depending on other fetchers states via stage).

The three methods of `FetchStrategy`:
```python
def fetch(self):
    pass

def check(self):
    pass

def expand(self):
    pass
```
have been merged in the single method:
```python
def fetch(self, validate=True, expanded_source_tree=any):
    pass
```
to simplify the logic in the code. Upstream classes followed (so e.g. `Stage.fetch` now has an `expand` argument and so does `PackageBase.do_stage`).

The convoluted logic in `PackageBase` to construct 2 composites (one for stages, one for fetchers) has been removed. As fetchers don't need a stage reference anymore, a `StageComposite` instance is now built in the simplest way possible (first the root stage, then the resources). This object can return on demand a `FetchStrategyComposite`.

For other minor details the commit messages should be of help.

##### Mandatory todo list

- [x] Decouple classes in `fetch_stretegy.py` from stages
- [x] Simplify the construction of composite objects in `PackageBase`
- [ ] Improve coverage for fetchers (refactor their unit tests?)
- [ ] Improve docstrings for fetchers and stages

##### Other changes on the wishlist

I refrained from doing a couple of changes that I thought should be discussed beforehand. The first one is a change in the `FetchStrategy.fetch` method which is not strictly needed to maintain the current behavior and pass tests. I would like the final signature to look like:
```python
    def fetch(self, validate=True, expanded_source_tree=any, archive=any):
        """Fetches the source code archive or repository.

        Args:
            validate (bool): if True, try to validate downloaded files whenever
                possible
            expanded_source_tree (bool or any): if True, the source tree is
                expanded, if False the source tree is granted not to be there.
                If 'any' it depends on the defaults of the derived class
            archive (bool or any): if True, the archive is present in the source
                directory. If false it's not. If 'any' it depends on the defaults of 
                the derived class

        Returns:
            absolute path of the archive file if any, else None

        Raises:
            FetchError: if anything goes wrong

        """
```
basically I would like to add an extra argument to drive the presence (or absence) or the archive. If implemented this will permit to ask for a specific state in the working directory, regardless of the fetcher we are using. I would do this more to have an interface that is not biased towards archive fetchers, rather then for an immediate need to implement some features.

The other change is just structural. I think `fetch_strategy.py` has enough code to be split into:
```console
fetch_strategy
├── vcs.py
├── url.py
├── __init__.py
└── exceptions.py
```
while maintaining the same interface to the outside.

Well, I'll be waiting for comments on the code!

### CODE REVIEW

File: lib/spack/spack/fetch_strategy.py [hunk 1]
```diff
@@ -40,89 +40,127 @@
...
+    _expected_archive_files = None
 
-    def __init__(self):
-        # The stage is initialized late, so that fetch strategies can be
-        # constructed at package construction time.  This is where things
-        # will be fetched.
-        self.stage = None
+    @property
+    def source_dir(self):
+        """Root directory where the source code should be expanded"""
```
Comment (alalazo) [inline lib/spack/spack/fetch_strategy.py]: @tgamblin For this I was quite undecided between `source_dir` and `working_dir`. Let me know if you have any opinion or a third name to put into the mix.

**Comment by alalazo** [comment]
Uh, what happened to `codecov`?

**Comment by tgamblin** [comment]
not sure -- i saw that on another PR as well.

**Comment by alalazo** [comment]
@tgamblin Any comment on the interface proposed above for `Fetcher.fetch`? Right now this refactoring is good enough to pass tests, and maintain the current behavior.

**Comment by alalazo** [comment]
ping

**Comment by alalazo** [comment]
@tgamblin I would love to finalize this one, if possible.

**Comment by alalazo** [comment]
Super ping: img1 **+** img2 ?

File: lib/spack/spack/fetch_strategy.py [hunk 2]
```diff
@@ -158,64 +194,65 @@ class FetchStrategyComposite(object):
...
-        self.url = kwargs.get('url', None)
-        if not self.url:
-            self.url = url
-
-        self.digest = next((kwargs[h] for h in crypto.hashes if h in kwargs),
-                           None)
+    def __init__(self, url, digest=None, **kwargs):
+        self.url = url
+        self.digest = next(
+            (kwargs[h] for h in crypto.hashes if h in kwargs), None
```
Comment (tgamblin) [inline lib/spack/spack/fetch_strategy.py]: I think the way I did this was bad (but definitely not the worst thing I did with fetchers -- so much of them is horrible).  There should be a failure here if they provide two digest types... also, we allow them to just pass the hash in without a name... there should probably be an error if they pass a `digest` argument _and_ a hash.  Ugh.

How about:
```python
digests = [] if digest is None else [digest]
digests += [kwargs[h] for h in kwargs if h in crypto.hashes]
if len(digests) > 1:
    raise ValueError('Fetchers can only use one digest')
self.digest = digests[0] if digests else None
```

File: lib/spack/spack/fetch_strategy.py [hunk 3]
```diff
@@ -158,64 +194,65 @@ class FetchStrategyComposite(object):
...
         if not self._curl:
             self._curl = which('curl', required=True)
         return self._curl
 
-    @_needs_stage
-    def fetch(self):
-        if self.archive_file:
-            tty.msg("Already downloaded %s" % self.archive_file)
-            return
+    def fetch(self, validate=True, expanded_source_tree=any):
```
Comment (tgamblin) [inline lib/spack/spack/fetch_strategy.py]: why does adding parameters like this simplify the logic?  I'm not really seeing it -- to me it makes the API harder to implement and less clearly divided into functions.  Does this help with the interaction between a fetcher and the stage somehow?

File: lib/spack/spack/fetch_strategy.py [hunk 4]
```diff
@@ -438,6 +474,8 @@ def fetch(self):
 
 class VCSFetchStrategy(FetchStrategy):
 
+    exclude_patterns = tuple()
```
Comment (tgamblin) [inline lib/spack/spack/fetch_strategy.py]: ()

File: lib/spack/spack/fetch_strategy.py [hunk 5]
```diff
@@ -55,74 +55,113 @@
...
+    """Registers a fetch strategy
+
+    Args:
+        cls (FetchStrategy): fetcher class to be registered
+    """
+    all_strategies.append(cls)
+    return cls
+
+
+def has_expanded_tree(directory):
```
Comment (tgamblin) [inline lib/spack/spack/fetch_strategy.py]: The `has_` prefix should be for things that actually return `bool`.  This doesn't -- I would call this `expanded_tree`.  It's much more legible to use that in a `bool` context than the other way around.

File: lib/spack/spack/fetch_strategy.py [hunk 6]
```diff
@@ -289,24 +325,28 @@ def fetch(self):
...
-        """Path to the source archive within this stage directory."""
-        return self.stage.archive_file
+        # Checksum validation
+        if validate:
+            self._check()
+
+        # Archive expansion
+        et = has_expanded_tree(self.source_dir)
+        do_expansion = expanded_source_tree in (any, True)
+        if do_expansion and not et:
```
Comment (tgamblin) [inline lib/spack/spack/fetch_strategy.py]: I find this whole if statement hard to follow -- also "do_expansion" reads awkwardly to me -- `expand` is a verb so why not just `expand`?

I would simplify it this way:

```python
if expanded_source_tree in (any, True):
    if has_expanded_tree(self.source_dir):
        tty.msg('Already expanded...')
    else:
        self._expand_archive(self.source_dir)
```

Now there are only two conditions to consider and not 2 variables + 2 cases with `and`/`and not`

File: lib/spack/spack/fetch_strategy.py [hunk 7]
```diff
@@ -55,74 +55,113 @@
...
+    #: Private cache for expected_archive_files,
+    #: initialized to None in the class
+    _expected_archive_files = None
 
-class FetchStrategy(with_metaclass(FSMeta, object)):
-    """Superclass of all fetch strategies."""
-    enabled = False  # Non-abstract subclasses should be enabled.
-    required_attributes = None  # Attributes required in version() args.
+    @property
+    def source_dir(self):
```
Comment (tgamblin) [inline lib/spack/spack/fetch_strategy.py]: I think we should rename this.  I find `source_dir` really confusing because it's actually the root of the stage here, not the expanded archive's source directory.  Why not `stage_path`?

I'm wondering whether this isn't still too entangled.  I think that maybe `stage` should be responsible for all directory management, and the fetcher should just assume that it is fetching into the current working directory all the time... That might simplify a lot of the logic here...  I need to think about this more.

File: lib/spack/spack/fetch_strategy.py [hunk 8]
```diff
@@ -55,74 +55,113 @@
...
 
+    #: Attributes required in version() args
+    required_attributes = tuple()
+    #: If set it will be used to save an archive to a specific filename
+    save_filename = None
+    #: Private cache for source_dir, initialized to None in the class
+    _source_dir = None
+    #: Private cache for expected_archive_files,
+    #: initialized to None in the class
+    _expected_archive_files = None
```
Comment (tgamblin) [inline lib/spack/spack/fetch_strategy.py]: I think this is still too entangled... we're saving parts of the stage in the fetcher and duplicating information now.  I think we really need to separate responsibilities more

File: lib/spack/spack/fetch_strategy.py (hunk 7 above)
Comment (alalazo) [reply lib/spack/spack/fetch_strategy.py]: Doesn't this mean that the client of a fetcher is responsible of moving to the correct directory, and then moving back when he's finished? I think it may simplify the logic here, not it's use within the code.

File: lib/spack/spack/fetch_strategy.py (hunk 2 above)
Comment (alalazo) [reply lib/spack/spack/fetch_strategy.py]: Part of the modifications here were along the lines of being defensive. I didn't want to pile up too many modifications on behavior without knowing why you did some of the things here and while I was modifying the architecture of stages and fetchers.

**Comment by alalazo** [comment]
> I haven't gone through everything yet b/c I have a more fundamental question: why is the new fetch() call simpler? I really like the disentanglement from the stage, but can't that happen without merging three very logically distinct functions into one? What's the win? Can we still disentangle without the API change?

My point of view is that the API is currently modeled on tarballs, and does not abstract correctly other fetchers. If you think of a VCS strategy, how can you say that `fetch`, `check` and `expand` are distinct operations? Checking the implementation they're all no-op:
```python
class VCSFetchStrategy(FetchStrategy):
    ...
    @_needs_stage
    def check(self):
        tty.msg("No checksum needed when fetching with %s" % self.name)

    @_needs_stage
    def expand(self):
        tty.debug("Source fetched with %s is already expanded." % self.name)
```
The fact that 4 out of 6 concrete fetchers were implementing interface methods as a no-op looked to me as a flaw in the API. What I did then was to go through all the calls of fetchers, and try to "distill" what we were using these 3 methods for. The common patterns I saw led to the API in the PR and to the proposed extension in the description. 

I also think the proposed interface makes code a lot more clear on where the operations are done. If you consider the snippet:
```python
def foo(fetcher):
    fetcher.fetch()
    # do something
    fetcher.check()
    # do something else
    fetcher.expand()
```
the expansion may now happen on `fetcher.fetch` or on `fetcher.expand`. The same is true for validation.
If instead you use the proposed API:
```python
def foo(fetcher):
    fetcher.fetch(validate=True, expanded_source_tree=True, archive=any)
    # ...
```
you clearly express your intents (I want to validate the fetch, have an expanded source tree, I don't care if a tarball is there or not) and you are ensured that on return from the call things are like you asked.
